PDAL: enable use of system libunwind, drop dependency of port

nilason opened the pull request:
#### Description

This doesn't solve the problem with Xcode 15, automatically linking to MacPorts' libunwind (if present) to all C++ based ports, but it makes it possible to some degree avoid installing the port in the first place.

See https://trac.macports.org/ticket/68250
See https://trac.macports.org/ticket/66250

The patch added with this PR is already committed to master upstreams.

This at least minimises the risk of "polluting" direct PDAL dependants qgis3, grass, saga with unwanted libunwind linking.


<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 13.6 22G120 arm64
Xcode 15.0 15A240d

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

barracuda156 commented:
@nilason Has this been tested on powerpc? There is no system libunwind there.

nilason commented:
> @nilason Has this been tested on powerpc? There is no system libunwind there.

libunwind support is an optional part of PDAL, so that shouldn’t in itself be an issue. The bots have however no good track record with pre-10.7 systems as it is.

i0ntempest commented:
I’m trying to address this in #20632 and I’ve made mp libunwind optional as a variant. If you want we could combine our solutions.

barracuda156 commented:
@nilason @i0ntempest Thank you for clarifying!

nilason commented:
> I’m trying to address this in #20632 and I’ve made mp libunwind optional as a variant. If you want we could combine our solutions.

Sorry, I overlooked that one. Now I see only the main maintainer is notified.

As I mentioned in https://trac.macports.org/ticket/66250 I don’t see much reason to keep the port libunwind for 10.7+ (or maybe even 10.6+) systems. It is after all part of the system. It seems to me as of Xcode 15 it is even a ticking time bomb. I’m not sure what to gain with a new variant for that, but maybe I’m missing something.

i0ntempest commented:
> > I’m trying to address this in #20632 and I’ve made mp libunwind optional as a variant. If you want we could combine our solutions.
> 
> Sorry, I overlooked that one. Now I see only the main maintainer is notified.
> 
> As I mentioned in https://trac.macports.org/ticket/66250 I don’t see much reason to keep the port libunwind for 10.7+ (or maybe even 10.6+) systems. It is after all part of the system. It seems to me as of Xcode 15 it is even a ticking time bomb. I’m not sure what to gain with a new variant for that, but maybe I’m missing something.

Good point. My change is less destructive (?) but if it’s not needed (and is in the system) I’m ok with just dropping the dependency. My variant can always be added in later (and I want to get rid of mp libunwind on all my systems asap).

barracuda156 commented:
@i0ntempest Given that there are systems without libunwind, it makes sense to provide a non-default variant.

i0ntempest commented:
@nilason Can your patched cmake file find both MP libunwind and the system one? If both are found which one will it use?
I'm thinking you merge this first and I'll update my PR to add the variant to link MP libunwind.

nilason commented:
Now I finally found the time to do some real testing, with and without PDAL "backtrace support". As far as I could tell, the PDAL backtrace code seems to be not needed for Mac OS. Either way, with or without, the system output a nice, similar looking backtrace (as may be conveniently viewed in the Console.app). I'm now quite convinced that the dependency of libunwind is absolutely unnecessary. And as the feature is optional and doesn't add functional features for using PDAL, I wouldn't bother to add this dependency to pre-10.6 systems (on which PDAL [nor libunwind, according to bot builds] do not work anyway).

I reworked this PR a bit. I added a new patch, which I submitted upstreams (still pending), it enables a configure option flag `-DWITH_BACKTRACE`.  This flag is set to OFF. I also set the port to `conflicts libunwind` for Xcode 15+.

PDAL is the only port in the GIS stack of ports that up to this point had the libunwind port as a dependency. Removing this dependency might solve this problem at least for this group of users.

barracuda156 commented:
> on which PDAL [nor libunwind, according to bot builds] do not work anyway

@nilason Are tests failing for PDAL or what is the issue particularly?

nilason commented:
> > on which PDAL [nor libunwind, according to bot builds] do not work anyway
> 
> @nilason Are tests failing for PDAL or what is the issue particularly?

The bots are failing according to https://ports.macports.org/port/PDAL/details/.

nilason commented:
> > > on which PDAL [nor libunwind, according to bot builds] do not work anyway
> > 
> > 
> > @nilason Are tests failing for PDAL or what is the issue particularly?
> 
> The bots are failing according to https://ports.macports.org/port/PDAL/details/.

(Looks like thread_local related issue.)

barracuda156 commented:
> > > > on which PDAL [nor libunwind, according to bot builds] do not work anyway
> > > 
> > > 
> > > @nilason Are tests failing for PDAL or what is the issue particularly?
> > 
> > 
> > The bots are failing according to https://ports.macports.org/port/PDAL/details/.
> 
> (Looks like thread_local related issue.)

Perhaps we need `compiler.thread_local_storage yes` being set, so that a correct compiler is chosen. On 10.6 it builds fine with `gcc12`.

kencu commented:
Rather than do this kind of thing for every port that will opportunistically pick up macports' libunwind port when installed, it would probably be much better to have the libunwind port just be a stub port on 10.7+.

Then all that is needed is to revbump all the things that have opportunistically linked with it in the past, and you're done.

Disabling unwinding support completely, or working around it port-by-port, is both a lot of unnecessary work and can reduce/disable features.

nilason commented:
> Rather than do this kind of thing for every port that will opportunistically pick up macports' libunwind port when installed, it would probably be much better to have the libunwind port just be a stub port on 10.7+.
> 
> Then all that is needed is to revbump all the things that have opportunistically linked with it in the past, and you're done.

Can't agree more. If you have read anything I have written, in this PR description and in the linked tickets, you'd understand I'm quite aware of the danger having MP's libunwind around with Xcode 15 and that this is not a general solution. However, I'm personally neither experienced nor resourced enough for that wide-ranging feat.

This PR is putting out the fire in one key corner in the GIS area, with an important port `qgis` completely broken on the newest systems that need to be addressed promptly and shouldn't be delayed.

i0ntempest commented:
I'll close my PR in favor of the more permanent solution proposed by Ken. meanwhile I think this should still be merged so QGIS can be usable before the permanent solution is implemented.

nilason commented:
> I'll close my PR in favor of the more permanent solution proposed by Ken. meanwhile I think this should still be merged so QGIS can be usable before the permanent solution is implemented.

Please, do merge.

nilason commented:
> Perhaps we need `compiler.thread_local_storage yes` being set, so that a correct compiler is chosen. On 10.6 it builds fine with `gcc12`.

@barracuda156 Thanks for testing, will add it with next version bump.


@i0ntempest Thanks!
